### PR TITLE
Raise a warning when `nn.Module` instance is saved with `save_hyperparameters()`

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Addresses: https://github.com/PyTorchLightning/pytorch-lightning/issues/11834#issuecomment-1034752642

`nn.Module`s are not hparams and when someone uses LM as systems, they might forget to ignore the arguments and they will be saved inside hparams thus will increase the size of the checkpoint. 

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: pytorch_lightning/utilities/parsing.py [hunk 1]
```diff
@@ -247,9 +248,15 @@ def save_hyperparameters(
             hp = {arg: init_args[arg] for arg in args if isinstance(arg, str)}
             obj._hparams_name = "kwargs"
 
+    hp = obj._to_hparams_dict(hp)
+    # pop all nn.Module
+    for key in list(hp):
+        if isinstance(hp[key], nn.Module):
+            del hp[key]
```
Comment (krshrimali) [inline pytorch_lightning/utilities/parsing.py:255]: https://github.com/PyTorchLightning/pytorch-lightning/blob/52013f066641f07c8abb4a27dc0583447101afb9/pytorch_lightning/utilities/parsing.py#L234

Can we just add a check if key is not a `nn.Module` instance in the line above and avoid deleting?

File: tests/models/test_hparams.py [hunk 2]
```diff
@@ -320,8 +320,7 @@ def test_collect_init_arguments(tmpdir, cls):
     if isinstance(model, (SubClassBoringModel, NonSavingSubClassBoringModel)):
         assert model.hparams.subclass_arg == 1200
 
-    if isinstance(model, AggSubClassBoringModel):
-        assert isinstance(model.hparams.my_loss, torch.nn.CosineEmbeddingLoss)
+    assert "my_loss" not in model.hparams
```
Comment (krshrimali) [inline tests/models/test_hparams.py:323]: Does this mean that the PR introduces a BC-breaking change?

**Review by krshrimali — CHANGES_REQUESTED** [review]

File: pytorch_lightning/utilities/parsing.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/parsing.py:255]: yep!

File: tests/models/test_hparams.py (hunk 2 above)
Comment (rohitgr7) [reply tests/models/test_hparams.py:323]: sort of yes.. I actually wanted to discuss this. since my main motivation was to disable saving the modules only, but losses are also inherited from `nn.Module`. If we agree with this, we can mark this as breaking change.

**Comment by awaelchli** [comment]
It's probably ok to do this, but we should stop here. There are many other parameter type that may ore may not be desired to serialize, however we can't do the guess work for everything. 

Note: If we go with this PR, there is no way for the user to force what is being ignored here. I hesitate to accept here, because normally we shouldn't make such assumptions for the user (the framework should not be limiting them).

**Comment by rohitgr7** [comment]
> It's probably ok to do this, but we should stop here. There are many other parameter type that may or may not be desired to serialize, however we can't do the guess work for everything.
> 
> Note: If we go with this PR, there is no way for the user to force what is being ignored here. I hesitate to accept here, because normally we shouldn't make such assumptions for the user (the framework should not be limiting them).

I agree, my idea here was to avoid only `nn.Modules` because that's already been captured when we save `state_dict` inside the checkpoint. Other different types of arguments will not be captured automatically.

I was running t5-small, and it got stuck due to this, with no errors at all, so it was hard to figure out what went wrong here until I did some debugging with @krshrimali .

We can mention this inside docstrings if required.

File: tests/models/test_hparams.py (hunk 2 above)
Comment (carmocca) [reply tests/models/test_hparams.py:323]: Yeah, that's a big limitation as the loss is often made configurable.

**Comment by carmocca** [comment]
I'm also hesitant about this.

> I was running t5-small, and it got stuck due to this, with no errors at all, so it was hard to figure out what went wrong here until I did some debugging with @krshrimali .

What if we focus instead on notifying the user about the associated problems? We can suggest excluding the keys with `save_hyperaprameters`

**Comment by rohitgr7** [comment]
> I'm also hesitant about this.
> 
> > I was running t5-small, and it got stuck due to this, with no errors at all, so it was hard to figure out what went wrong here until I did some debugging with @krshrimali .
> 
> What if we focus instead on notifying the user about the associated problems? We can suggest excluding the keys with `save_hyperaprameters`

well, that's a good suggestion. works for me since it will happen during init. If everyone agrees on this, I'll update the PR :)

File: pytorch_lightning/utilities/parsing.py [hunk 3]
```diff
@@ -225,12 +226,23 @@ def save_hyperparameters(
...
+    for k in list(init_args):
+        if k in ignore:
+            del init_args[k]
+            continue
+
+        if isinstance(init_args[k], nn.Module):
+            rank_zero_warn(
+                f"Attribute {k!r} is an instance of `nn.Module` and is already saved during checkpointing."
+                " It is recommended to ignore them using `self.save_hyperparameters(ignore=[k!r])`."
+            )
```
Comment (carmocca) [inline pytorch_lightning/utilities/parsing.py:245]: I'd prefer

```python
modules = [arg for arg in init_args if arg not in ignore and isinstance(arg, nn.Module)]
for module in modules:
    warn(...)
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/parsing.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/parsing.py:245]: I am doing 2 things here. Filtering the ones that are required to be ignored and raising a warning. can't seem to find a way to do both in a single loop.

File: pytorch_lightning/utilities/parsing.py [hunk 4]
```diff
@@ -247,9 +259,10 @@ def save_hyperparameters(
             hp = {arg: init_args[arg] for arg in args if isinstance(arg, str)}
             obj._hparams_name = "kwargs"
 
+    hp = obj._to_hparams_dict(hp)
```
Comment (rohitgr7) [inline pytorch_lightning/utilities/parsing.py:262]:
```suggestion
```

File: pytorch_lightning/utilities/parsing.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/utilities/parsing.py:245]: But what's the issue with doing it in 2 loops as in my suggestion?

Comment (krshrimali) [reply pytorch_lightning/utilities/parsing.py:245]: > But what's the issue with doing it in 2 loops as in my suggestion?

Probably, `del init_args[k]` is also required (and hence not a list comprehension)?

**Review by krshrimali — APPROVED** [review]
LGTM, thanks Rohit!

File: pytorch_lightning/utilities/parsing.py [hunk 5]
```diff
@@ -225,12 +226,23 @@ def save_hyperparameters(
...
+        ignore = {ignore}
+    elif isinstance(ignore, (list, tuple)):
+        ignore = {arg for arg in ignore if isinstance(arg, str)}
+
+    for k in list(init_args):
+        if k in ignore:
+            del init_args[k]
+            continue
+
+        if isinstance(init_args[k], nn.Module):
```
Comment (Borda) [inline pytorch_lightning/utilities/parsing.py:241]: shall we also add an argument that the user would need to explicitly allow saving `nn.Module` otherwise drop it?

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/utilities/parsing.py (hunk 5 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/parsing.py:241]: this will save the `nn.Modules`, but will raise a warning in that case, because of checkpointing issue. I don't think we need an argument since that will only disable the warnings I guess.
